docs/design_docs/user_facing_api.md

```diff
@@ -0,0 +1,198 @@
+# Proposal for user facing API
+
+## Objectives
+- deploy the plugins without day 1 configuration
+- dynamic day 2 host resource exposure
+- expose lower device as a logical resource
+- be able to use node selectors / node affinity / NIC selectors /
+resource name
+
+## Requirements
+- keep the DP as independent from multus and CNI as possible
+- keep the explicit configuration (configMap)
+- have an implicit configuration (would be overriden by the explicit)
+
+## Alternatives
+
+### Expose all host interfaces
+This approach would be valid for MVP.
+
+It would:
+  - read data from configMap
+  - if configMap provides empty configuration
+    - read all devices & bonds
+    - expose 1 resource per deviceOrBond with hard-coded values for:
+      - resource name: it would use the 'device' name
+      - mode: bridge
+      - capacity: 100
+
+This approach would fulfil 3/5 objectives, missing:
+- expose lower device as a logical resource
+- be able to use node selectors / node affinity / NIC selectors /
+resource name
+
+This approach is implemented in
+[PR #10](https://github.com/kubevirt/macvtap-cni/pull/10), and is considered
+good enough for the MVP.
+
+### Host device exposure using selectors
+This proposal complements the proposal where
+[all interfaces are exposed](#expose-all-host-interfaces). It would follow the
+same behavior of
+[k8s network policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/#isolated-and-non-isolated-pods);
+by default all host's interfaces are exposed, but once an host is selected via
+node selector, a white-list mechanism is triggered.
+
+It would consist of 3 stages:
+  - tagging the hosts
+  - provisioning the macvtap policy
+  - provisioning the macvtap NAD
+
+#### Host tagging
+Host tagging could be based on:
+  - macvtapAware: bool (indicates if it is selectable via node selector)
+  - !node-role.kubernetes.io/master
+  - tag the nodes with the interface names you want it to expose (???)
+    - e.g. parent_iface.macvtap.network.kubevirt.io/eth0
+
+#### Macvtap policy
+The macvtap policy would look like this, should we choose a node-centric
+approach:
+```yaml
+apiVersion: ...
+kind: MacvtapNetworkNodePolicy
+metadata:
+  name: policy-1
+  namespace: macvtap-cni
+spec:
+  nicSelector:
+    - parentIface: bond1
+      mode: private
+      quantity: 25
+      resourceName: bondedIface
+    - parentIface: eth1
+      mode: vepa
+      resourceName: quickDataplane
+    - parentIface: eth2
+      resourceName: dataplane
+  nodeSelector:
+    matchLabels:
+      macvtapAware: true
+```
+For each node matching the `nodeSelector` knob, it would expose all interfaces
+on that host that are featured in the `nicSelector` list. Each resource would
+be advertised under the name specified in the `resourceName` attribute.
```

Comment:
How would this work if multiple NICs match the policy on a node? Which one would be passed to CNI? I think we need to ensure that only one parent iface can be exposed per node. We can pick the first matching selector in the list.